iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/error_kind.hpp

```diff
@@ -0,0 +1,130 @@
+#ifndef IOX_HOOFS_ERROR_REPORTING_ERROR_KIND_HPP
+#define IOX_HOOFS_ERROR_REPORTING_ERROR_KIND_HPP
+
+#include <cstdint>
+#include <type_traits>
+
+namespace iox
+{
+namespace err
+{
+using error_level_t = uint32_t;
+constexpr error_level_t FATAL_LEVEL{0};
+
+// mandatory fatal error category that always exists
+struct Fatal
+{
+    static constexpr char const* name = "Fatal";
+
+    static constexpr error_level_t value = FATAL_LEVEL;
+
+    explicit operator error_level_t()
+    {
+        return value;
+    }
+};
+
+struct PreconditionViolation
+{
+    static constexpr char const* name = "PreconditionViolation";
+
+    static constexpr error_level_t value = FATAL_LEVEL;
+
+    explicit operator error_level_t()
+    {
+        return value;
+    }
+};
+
+// postconditions and other asserts, not for preconditions
+struct DebugAssertViolation
+{
+    static constexpr char const* name = "DebugAssertViolation";
+
+    static constexpr error_level_t value = FATAL_LEVEL;
+
+    explicit operator error_level_t()
+    {
+        return value;
+    }
+};
+
+template <class T>
+struct IsFatal : public std::false_type
+{
+};
+
+template <>
+struct IsFatal<Fatal> : public std::true_type
+{
+};
+
+template <>
+struct IsFatal<PreconditionViolation> : public std::true_type
+{
+};
+
+template <>
+struct IsFatal<DebugAssertViolation> : public std::true_type
+{
+};
+
+template <class Kind>
+bool constexpr isFatal(Kind)
+{
+    return false;
+}
+
+template <>
+bool constexpr isFatal<Fatal>(Fatal)
+{
+    return true;
+}
+
+template <>
+bool constexpr isFatal<PreconditionViolation>(PreconditionViolation)
+{
+    return true;
+}
+
+template <>
+bool constexpr isFatal<DebugAssertViolation>(DebugAssertViolation)
+{
+    return true;
+}
+
+template <typename Kind>
+bool constexpr requiresHandling(Kind)
+{
+    return true;
```

Comment:
No, the intention is that the primary template assumes it has to be handled, unless explicitly specified otherwise. It is a opt in vs. opt out question.

Reply:
Removed this trait entirely.